Hexagons moved to own component

rschloen opened the pull request:
Created new component, Hexagons, to hold actuator layout so that it can be used more easily in other components(Manual, Automatic). 

At the moment, Hexagon component dispatches an event when an actuator is clicked which calls functions in Manual to generate the actuator commands. This will likely be changed to make actuator control more global rather than repeating the functions that generate the command in every parent component. (Note: Automatic component does not currently have functions that build commands in anticipation of new state tracking scheme).

Created new orientation for the hexagon layout: vertical. Orientation refers to the direction in which the hexagons are in line with their neighbors. For example, actuators 0-3 line up in a column in the vertical orientation, and a row in horizontal.

rschloen committed:
Moved Hexagons into separate component

To make it easier to use the hexagons in different pages, I extracted
the hexagon creation code and made a new Hexagon component that can be
imported into other parent components like Manual.

Added event dispatchers to Hexagon component that call functions in
parent component (ei: Manual) like buildCommandBlock for normal use and
testing.

Used bind:prop to communicate changes to props, like activeHexagon and
test_ui, in the parent component to Hexagon component.

rschloen committed:
Changed "specify" route to be default ("/"") route

rschloen committed:
Added scrollbox with linked options to Automatic

Scrollbox contains several placeholder options that are linked to
different configurations such as different orientations for the actuators
or presets.

rschloen committed:
Added new orientation for hexagon layout

Parent component can set orientation of hexagon layout (horizontal/
vertical).

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,267 @@
+const dispatch = createEventDispatcher();
+export let test_ui = false;
+export let activeHexagon = -1;
+export let orientation = "horizontal"; //orientation refers to direction numbers are listed (horizontal:left to right, vertical:top to bottom)
```
In the future, I would like to take a hexagon config as a props and some intermediate components would take in orientation or hexagon config ids and spit out the horizontalHexagons array and transformation string

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,267 @@
+  // filler
+  { id: 34, row: 3, col: 6 },
+  { id: 35, row: 1, col: 6 },];
+  const hexagonSideLength = 70;
```
spacing is weird here

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,267 @@
+  { id: 34, row: 3, col: 6 },
+  { id: 35, row: 1, col: 6 },];
+  const hexagonSideLength = 70;
+const globalPadding = 10;
+const baseSpacing = 15;
```
This is the bit that should be computed to a degree to avoid clipping depending on where things are placed. Let's worry about fixing this once we have more stuff in place. We will have to account for it if we are going to overlay the  fabric on the body placement images in the scrollable options. ... Maybe we can skip handling the spacing math and get the svg scaling to handle it for us in places where we don't set touch input handlers.

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,267 @@
+        x: Math.max(previousValue.x, currentValue.x + (height / 2) + (globalPadding / 2)),
+        y: Math.max(previousValue.y, currentValue.y + (width / 2) + (globalPadding / 2)),
+    };
+});
```
I don't like how all of this is duplicated. This seems like it will be hard to translate to stichable fabrics. What is your plan?

trueb2 reviewed (changes requested):
From some of the stuff that Yei Hwan showed me, I think we need to use translation math to turn acutator blocks instead of if statements. If we go diagonal, then this gets very hard to check.

If we add the math for rotation to the hit check to the handleTouchMove, then we only need to change the rotation on the svg. All of the ids, padding, and centers will still be correct

rschloen committed:
Switched to using svelte-moveable

svelte-moveable replaces math that changes hexagon layout based on
orientation, and allows the the container with the hexagon svg to
rotate.

Math inside handleTouchMove is not changed, and touch offset values stay
 relative to the rotated div container.

Container can be rotated with cursor or by button which calls a method
of Moveable to rotate by a number of degrees or rotate to an angle.

rschloen committed:
Added buttons for rotation control

Added an input where the user can set degrees to rotate, and buttons to
rotate the hexagons the set degrees. Add button to reset hexagons to
initial orientation.

rschloen committed:
Changed formatting/layout of Hexagons

Fixed layout in Manual page so hexagons don't get cut off an need
scroll bar. Changed font-size of buttons next to hexagons.

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -268,6 +269,7 @@ function handleActuatorClick(e) {
                 {/each}
             </ul>
         </div><div class='col-25'>
+            <br/><br/><br/>
```
🤔

rschloen replied:
That was just a temporary 'hacky' way to fix the spacing so the timing column lined up with messages at the top of the column on the left when I changed the layout so the hexagons don't get cutoff at the bottom. Will likely be replaced when I go through and finalize the layout for everything.

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,237 @@
+
+const dispatch = createEventDispatcher();
+
+let target;
+const frame = {rotate: 0,};
+let moveable;
+
+export let test_ui = false;
+export let activeHexagon = -1;
+
```
style-wise, I would put exports (external immutable globals), then lets (mutable globals), then consts (immutable globals).

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,237 @@
+                    stroke={hexagon.color}
+                    stroke-width="5px"
+                    strokeLinejoin="miter"
+                    transform={`rotate(0 ${hexagon.x} ${hexagon.y}) translate(${hexagon.x - (hexagon.width / 2)} ${hexagon.y - (hexagon.height / 2)})`}
```
```suggestion
                    transform={`translate(${hexagon.x - (hexagon.width / 2)} ${hexagon.y - (hexagon.height / 2)})`}
```

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -0,0 +1,237 @@
+        target.style.transform = `rotate(${beforeRotate}deg)`;
+    }}
+on:rotateEnd={({ detail: { target, isDrag, clientX, clientY }}) => {
+    console.log("onRotateEnd", target, isDrag);
```
This log was probably useful for debugging but won't be in the future. I would recommend deleting.

trueb2 commented on src/routes/Automatic.svelte:
```diff
@@ -1,7 +1,72 @@
+<style>
+    .scroll-box{
+        width: 25%;
+        height: 750px;
```
Please don't use pixels where possible. We want to be using em's so that the iPad, Surface, and laptop browser experiences stay similar. At this point I think that we are going to have a final sweep at the end to go through and fix our px usages.

trueb2 commented on src/routes/Automatic.svelte:
```diff
@@ -1,7 +1,72 @@
+    .option{
+        width: 100%;
+        height: 300px;
+        /* margin: 5px; */
```
dead code :: code smell. I know this about to change so no biggie

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -129,9 +129,7 @@ function setTimingBlock(config) {
-$: pendingTimeout = false;
+let activeHexagon = -1;
 
 let block0_31 = [0,0,0,0];
```
If we could pull out the block math and command building, this file could start to be really clean. Pulling this apart further would be overkill. I think this is just about the right amount of complexity until there are further needs.

trueb2 reviewed (commented):
I left some comments. None of them are blockers. I like these changes. Since there are more to come in a similar vein for the Automatic page, I will leave it up to you. Do you want me to merge this, while you handle cleanup in a separate PR or would you rather put a bow on these changes before merging?

rschloen committed:
Cleaned up code before merge

Deleted dead code, reorganized variable initialization, Deleted unneeded
logs.
